### PR TITLE
[py] fix: SecurityError: Invalid Domain problem

### PR DESCRIPTION
**Pull request by SinghHrmn**
### Description

- Fixed the Invalid Domain error while executing JS in Virtual
Authenticator tests.
- Added more tests for Virtual Authenticator.
- Updated SimpleWebserver to use 'localhost' when serving a page when `localhost` check is `True`. 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: py/test/selenium/webdriver/common/webserver.py [hunk 1]
```diff
@@ -163,7 +164,10 @@ def stop(self):
         with contextlib.suppress(IOError):
             _ = urllib_request.urlopen(f"http://{self.host}:{self.port}")
 
-    def where_is(self, path) -> str:
+    def where_is(self, path, localhost=False) -> str:
+        # True force serve the page from localhost
+        if localhost:
+            return f"http://localhost:{self.port}/{path}"
```
Comment (AutomatedTester) [inline py/test/selenium/webdriver/common/webserver.py]:
```suggestion
            return f"http://{DEFAULT_HOST}:{self.port}/{path}"
```

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py [hunk 2]
```diff
@@ -144,109 +151,246 @@ def test_add_and_remove_non_resident_credentials(driver, pages):
...
+    for credential in credentials:
+        # Using startswith because there can be padding difference '==' or '=' in the end
+        if credential.id.startswith(extract_id(response1)):
+            credential1: Credential = credential
+        elif credential.id.startswith(extract_id(response2)):
+            credential2: Credential = credential
+        else:
+            assert False, "Unknown credential"
+
+    if credential1.is_resident_credential:
```
Comment (AutomatedTester) [inline py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: What happens if these are not True? Should the test fail?

Comment (AutomatedTester) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: same question for below where we have `assert True`

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py [hunk 3]
```diff
@@ -144,109 +151,246 @@ def test_add_and_remove_non_resident_credentials(driver, pages):
...
 
-    login = driver.find_element(By.ID, "login-button")
-    WebDriverWait(driver, 50).until(lambda x: x.find_element(By.ID, "login-button").is_displayed())
-    login.click()
+    driver.get(pages.url("virtual-authenticator.html", localhost=True))
 
-    WebDriverWait(driver, 40).until(lambda x: x.find_element(By.CLASS_NAME, "col-lg-12").is_displayed())
+    # Register 2 credentials
+    response1 = driver.execute_async_script(
+        "registerCredential().then(arguments[arguments.length - 1]);"
```
Comment (AutomatedTester) [inline py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: let's move this JS to a variable and then we can simplify the code here and below. It means any future changes only need to happen in 1 place

**Review by AutomatedTester — CHANGES_REQUESTED** [review]

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py (hunk 2 above)
Comment (SinghHrmn) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: yes it should fail. I'll add the assert False in the else case. Thanks for this I formatted code according to flake8 forget to add else block

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py (hunk 3 above)
Comment (SinghHrmn) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: Sure

**Comment by SinghHrmn** [comment]
@AutomatedTester @pujagani Code Review changes Done!

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py [hunk 4]
```diff
@@ -144,109 +152,240 @@ def test_add_and_remove_non_resident_credentials(driver, pages):
...
+    credential = Credential.create_resident_credential(
         bytearray({1, 2, 3, 4}),
         "localhost",
-        b64decode(BASE64__ENCODED_PK),
+        bytearray({1}),
+        urlsafe_b64decode(base64_pk),
         0,
     )
+    try:
+        driver.add_credential(credential)
```
Comment (AutomatedTester) [inline py/test/selenium/webdriver/common/virtual_authenticator_tests.py:202]: we should use [`pytest.raises`](https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions)

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py [hunk 5]
```diff
@@ -144,109 +152,240 @@ def test_add_and_remove_non_resident_credentials(driver, pages):
...
+    for credential in credentials:
+        # Using startswith because there can be padding difference '==' or '=' in the end
+        if credential.id.startswith(extract_id(response1)):
+            credential1: Credential = credential
+        elif credential.id.startswith(extract_id(response2)):
+            credential2: Credential = credential
+        else:
+            assert False, "Unknown credential"
+
+    if credential1.is_resident_credential:
```
Comment (AutomatedTester) [inline py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: What I meant is that we should have something like

```suggestion
    assert credential1.is_resident_credential == True, "Credential1 should be resident credential"
```

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py [hunk 6]
```diff
@@ -144,109 +152,240 @@ def test_add_and_remove_non_resident_credentials(driver, pages):
...
+        assert True
+    else:
+        assert False, "Credential1 should be resident credential"
+    if credential1.private_key is not None:
```
Comment (AutomatedTester) [inline py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: This would be 
```suggestion
    assert credential1.private_key is not None
```

**Review by AutomatedTester — CHANGES_REQUESTED** [review]
Update your asserts as I mentioned in some of the comments to make it look neater

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py (hunk 5 above)
Comment (SinghHrmn) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: I was writing the tests like this. But Flake8 complains about writing tests like this. 
for code 
```python
assert credential1.is_resident_credential == True, "Credential1 should be resident credential"
```
Error
```json
[{
	"resource": "/Users/harmandeepsingh/BrowserStack/selenium/py/test/selenium/webdriver/common/virtual_authenticator_tests.py",
	"owner": "python",
	"code": "E712",
	"severity": 8,
	"message": "comparison to True should be 'if cond is True:' or 'if cond:'",
	"source": "flake8",
	"startLineNumber": 242,
	"startColumn": 47,
	"endLineNumber": 242,
	"endColumn": 47
}]
```

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py (hunk 6 above)
Comment (SinghHrmn) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: Same as above

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py (hunk 5 above)
Comment (SinghHrmn) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: Maybe we can skip flake8 in tests to clean the code.

Comment (SinghHrmn) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: @AutomatedTester comparing it with `==` has flake8 problem. But if I compare with `is` then flake8 is happy. So I've changed accordingly.

Comment (symonk) [reply py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: I don't like the idea of skipping flake8 in tests, booleans are singleton values and should be compared with `is` on identity (same with `None`) if not using the built-in implicit truthy checks.  You can just `assert credential1.is_resident_credential` here, checking the value is true is implicit

File: py/test/selenium/webdriver/common/virtual_authenticator_tests.py [hunk 7]
```diff
@@ -88,15 +94,36 @@ def create_rk_disabled_authenticator(driver) -> WebDriver:
...
+def extract_id(response):
+    return response.get("credential", {}).get("id", "")
+
+
+def extract_raw_id(response):
+    return response.get("credential", {}).get("rawId", "")
+
+
+def not_allowed_error_in(response) -> bool:
+    if response.get("status", "").startswith("NotAllowedError"):
```
Comment (symonk) [inline py/test/selenium/webdriver/common/virtual_authenticator_tests.py]: can just return the `response.get(...)` call here without the `if`

**Review by AutomatedTester — APPROVED** [review]

**Pull request merged**